compiler/src/iree/compiler/Dialect/HAL/Conversion/StreamToHAL/test/cmd_ops.mlir

```diff
@@ -203,9 +221,19 @@ func.func @cmdDispatch(%arg0: !stream.resource<transient>, %arg1: index, %arg2:
   %c128 = arith.constant 128 : index
   // CHECK: %[[CMD:.+]] = hal.command_buffer.create
   %0 = stream.cmd.execute with(%arg0 as %arg4: !stream.resource<transient>{%arg1}, %arg2 as %arg5: !stream.resource<external>{%arg3}) {
-    // Switch for each executable variant:
-    // CHECK: hal.device.switch
-    // CHECK-NEXT: #hal.device.match.executable.format<"embedded-elf-x86_64">
+    // Switch for each executable variant by checking conditions and ranking:
+    // CHECK: %[[DEVICE:.+]] = hal.command_buffer.device<%[[CMD]] : !hal.command_buffer>
+    //  CHECK-DAG: %{{.+}}, %[[AARCH64_FORMAT:.+]] = hal.device.query<%[[DEVICE]] : !hal.device> key("hal.executable.format" :: "embedded-elf-aarch64")
+    //  CHECK-DAG: %[[AARCH64_FEATURE:.+]] = scf.execute_region -> i1 {
+    // CHECK-NEXT:   %{{.+}}, %[[FEATURE:.+]] = hal.device.query<%[[DEVICE]] : !hal.device> key("some" :: "feature")
+    // CHECK-NEXT:   scf.yield %[[FEATURE]]
+    // CHECK-NEXT: }
```

Comment:
Around how many executable variants are we expecting to produce and then select from? I'm wondering about execution time and binary size overheads, and what other techniques for selecting between code paths could be used instead.

Seems great to have this either way, but I'm not quite seeing through all the layers for how heterogeneous execution / runtime feature selection / etc. will use these pieces.

Reply:
Depends on the config - hopefully one per major CPU arch a user is targeting and one per major unique set of extensions in SPIR-V (physical addressing or not, etc). We shouldn't be exposing multiple for every single little feature and should instead use executable constants (specialization constants in SPIR-V) wherever possible. So for a vmfb targeting CPUs maybe 1-4, for Vulkan 1-4, then combined may be 2-8, etc.